File: SortVision/public/code/selection/php/selectionSort.php
<?php
// Selection Sort Algorithm in PHP
//
// Selection Sort is a simple comparison-based sorting algorithm.
// It divides the array into a sorted and an unsorted region, and repeatedly selects
// the smallest (or largest) element from the unsorted region and moves it to the end of the sorted region.
// Time Complexity: O(n^2)
// Space Complexity: O(1) (in-place)

// Finds the index of the minimum element in the subarray from $start to $end
function findMinIndex($arr, $start, $end) {
    $minIndex = $start; // Assume the first element is the minimum
    for ($i = $start + 1; $i <= $end; $i++) {
        // Compare current element with current minimum
        if ($arr[$i] < $arr[$minIndex]) {
            $minIndex = $i; // Update minIndex if a smaller element is found
        }
    }
    return $minIndex; // Return the index of the minimum element
}

// Sorts the array in place using selection sort
function selectionSort(&$arr) {
    $n = count($arr); // Get the number of elements in the array
    for ($i = 0; $i < $n - 1; $i++) {
        // Find the index of the minimum element in the unsorted part
        $minIndex = findMinIndex($arr, $i, $n - 1);
        // Swap the found minimum element with the first element of the unsorted part
        if ($minIndex != $i) {
            $temp = $arr[$i];
            $arr[$i] = $arr[$minIndex];
            $arr[$minIndex] = $temp;
        }
    }
}

// -------------------
// Test Cases
// -------------------

// Test 1: Basic integer array
$arr1 = [64, 25, 12, 22, 11];
selectionSort($arr1);
echo "Sorted array 1: ";
print_r($arr1);

// Test 2: Array with negative numbers
$arr2 = [3, -1, 0, -7, 5, 2];
selectionSort($arr2);
echo "Sorted array 2: ";
print_r($arr2);

// Test 3: Already sorted array
$arr3 = [1, 2, 3, 4, 5];
selectionSort($arr3);
echo "Sorted array 3: ";
print_r($arr3);

// Test 4: Array with duplicates
$arr4 = [4, 2, 2, 8, 3, 3, 1];
selectionSort($arr4);
echo "Sorted array 4: ";
print_r($arr4);

// Test 5: Single element array
$arr5 = [42];
selectionSort($arr5);
echo "Sorted array 5: ";
print_r($arr5);

// Test 6: Empty array
$arr6 = [];
selectionSort($arr6);
echo "Sorted array 6: ";
print_r($arr6);
?>

File: SortVision/public/code/selection/scala/selectionSort.scala
/*
Selection Sort Algorithm in Scala 

What this file includes :-

-Implemented Selection Sort Algorithm
-Comments to make code clear
-Time and Space Complexity Analysis
-Example usage
-Test cases
-Performance Optimization Notes

Selection Sort Algorithm - Sorts the array by selecting a smallest integer and arraging it in ascending order.

*/

//Implementation

//Finding Minimum number's Index
object SelectionSort{
  def findMinIndex(arr: Array[Int], start: Int, end: Int): Int = {
    //assuming first element as minimum at start
    var minIdx = start

    //updating the minIdx
    for(i <- start+1 to end) {
      if(arr(i) < arr(minIdx)) {
        minIdx = i
      }
    }

    minIdx
  }

  //Sorting the array
  def selectionSort(arr: Array[Int]): Array[Int] = {

    val n = arr.length


    //arraging the elements at their appropriate place
    for(i <- 0 until n-1) {
      val minIdx = findMinIndex(arr, i, n-1)

      val temp = arr(i)
      arr(i) = arr(minIdx)
      arr(minIdx) = temp
    }

    arr
  }
}

/*

-Time Complexity - O(n^2)
-Space Complexity - O(1) -> Because of inplace substitution

*/


object SelectionSortTest {
  def main(args: Array[String]): Unit = {
     // Example Case
    val arr1 = Array(64, 34, 25, 12, 22, 11, 90)
    println("Original Array: " + arr1.mkString(", "))
    val sorted1 = SelectionSort.selectionSort(arr1.clone())
    println("Sorted Array:   " + sorted1.mkString(", "))


    //Test Cases
    // Test Case 1: Already sorted
    val arr2 = Array(1, 2, 3, 4, 5)
    assert(SelectionSort.selectionSort(arr2.clone()).sameElements(Array(1, 2, 3, 4, 5)))

    // Test Case 2: Reverse order
    val arr3 = Array(5, 4, 3, 2, 1)
    assert(SelectionSort.selectionSort(arr3.clone()).sameElements(Array(1, 2, 3, 4, 5)))

    // Test Case 3: All elements equal
    val arr4 = Array(7, 7, 7, 7)
    assert(SelectionSort.selectionSort(arr4.clone()).sameElements(Array(7, 7, 7, 7)))

    // Test Case 4: Empty array
    val arr5 = Array[Int]()
    assert(SelectionSort.selectionSort(arr5.clone()).isEmpty)

    // Test Case 5: Single Element
    val arr6 = Array[Int](1)
    assert(SelectionSort.selectionSort(arr6.clone()).sameElements(Array(1)))

    println("All tests passed!")
  }
}

/* Performace Optimization Notes */ 

/* 
-Not stable by default
- Good for small datasets
- Minimizes number of swaps compared to Bubble Sort
*/